src/he/pmtud.c

```diff
@@ -0,0 +1,287 @@
+  probe->length = payload_size;
+
+  conn->pmtud_probe_count++;
+  conn->pmtud_probing_size = probe_mtu;
+
+  // Send it
+  he_return_code_t res = he_internal_send_message(conn, buf, sizeof(he_msg_ping_t) + payload_size);
+  int timeout_ms = PMTUD_PROBE_TIMEOUT_MS;
+  if(res != HE_SUCCESS) {
+    // If he_internal_send_message failed to send the probe message, trigger the pmtud timeout with
+    // a tiny delay
+    timeout_ms = 10;
+  }
+
+  conn->pmtud_probe_pending_id = id;
+
+  // Start the probe timer
+  if(conn->pmtud_time_cb) {
+    conn->pmtud_time_cb(conn, timeout_ms, conn->data);
+  }
+
+  return HE_SUCCESS;
+}
+
+he_return_code_t he_internal_pmtud_handle_probe_ack(he_conn_t *conn, uint16_t probe_id) {
+  if(!conn) {
+    return HE_ERR_NULL_POINTER;
+  }
+  if(conn->pmtud_probe_pending_id != probe_id) {
+    // Receives an outdated probe id, ignore.
+    return HE_SUCCESS;
+  }
+
+  // Reset the pending probe id
+  conn->pmtud_probe_pending_id = 0;
+
+  // Reset the probe count for each ack
+  conn->pmtud_probe_count = 0;
+
+  // Handle the ack based on current state
+  switch(conn->pmtud_state) {
+    case HE_PMTUD_STATE_BASE:
+      // Base MTU confirmed
+      return he_internal_pmtud_base_confirmed(conn);
+    case HE_PMTUD_STATE_SEARCHING:
+      // Current probe acked
+      if(conn->pmtud_probing_size >= MAX_PLPMTU) {
+        // Search completed if the MAX_PLPMTU acked
+        return he_internal_pmtud_search_completed(conn);
+      } else {
+        // Increment probing size and send next probe
+        uint16_t probe_size = conn->pmtud_probing_size;
+        probe_size +=
+            (conn->pmtud_is_using_big_step) ? PMTUD_PROBE_BIG_STEP : PMTUD_PROBE_SMALL_STEP;
+        conn->pmtud_probing_size = MIN(probe_size, MAX_PLPMTU);
+        he_internal_pmtud_send_probe(conn, conn->pmtud_probing_size);
+      }
+      break;
+    case HE_PMTUD_STATE_ERROR:
+      // Base PMTU probe acked, enter Searching state now
+      return he_internal_pmtud_base_confirmed(conn);
+    default:
+      // Do nothing
+      break;
+  }
+  return HE_SUCCESS;
+}
+
+he_return_code_t he_internal_pmtud_handle_probe_timeout(he_conn_t *conn) {
+  if(!conn) {
+    return HE_ERR_NULL_POINTER;
+  }
+
+  // Try again with the same probe size
+  if(conn->pmtud_probe_count < MAX_PROBES) {
+    return he_internal_pmtud_send_probe(conn, conn->pmtud_probing_size);
+  }
+
+  // PROBE_COUNT reaches MAX_PROBES, decide what to do based on state
+  switch(conn->pmtud_state) {
+    case HE_PMTUD_STATE_BASE:
+      // Unable to confirm the base PMTU, entering error state.
+      return he_internal_pmtud_confirm_base_failed(conn);
+    case HE_PMTUD_STATE_SEARCHING:
+      if(conn->pmtud_is_using_big_step) {
+        // Try probing with small step
+        conn->pmtud_is_using_big_step = false;
+        conn->pmtud_probing_size -= PMTUD_PROBE_BIG_STEP;
+        conn->pmtud_probing_size += PMTUD_PROBE_SMALL_STEP;
```

Comment:
Same as before, think we don't need to update here and do it in `he_internal_pmtud_send_probe`
This reduces the number of places we update the value, which is easier to visualize

Reply:
Fixed.